Adds Lat/Lon option to categories

jackweinbender opened the pull request:
Issue #1061
  * Adds Lat/Lon to subject/article table
    * `DECIMAL` type to allow for specific decimal precision
    * Lat: validates between -90/90; Lon validates to -180/180; null allowed
    * Includes `gis_enabled?` method to determine if any related categories are GIS enabled
    * Conditionally displays Lat/Lon fields on subject/edit view based on `gis_enabled?`
    * Fixed form validation error on subject/edit form
  * Adds `gis_enabled` to category
    * Defaults to `false`
    * Enabling/Disabling GIS through "actions" menu affects all descendant nodes in tree
    * Changed "rename category" to "edit category" and added "Enable GIS" setting
    * Added "Enable GIS" to create category form
  * Adds URI field to all subjects

benwbrum commented on app/controllers/category_controller.rb:
```diff
@@ -36,8 +39,22 @@ def create
+    @category.descendants.each {|d| d.update_attribute(:gis_enabled, true)}
+    
+    flash[:notice] = "GIS Enabled for #{@category.title}"
+    ajax_redirect_to "#{request.env['HTTP_REFERER']}#category-#{@category.id}"
```
Path generation by mucking about with `HTTP_REFERER` is pretty ugly.  Do we already have a category_path URL helper we can use?  If not, it would be worth making one.  (We might need to pass in the controller as well.)

jackweinbender replied:
This whole controller seemed to use that method, so I went with it, assuming that there was something about it that I didn't know. My guess was that it had something to do with redirecting to specific anchors, though, that is perfectly doable with path helpers. I'll fix them all.

benwbrum commented on app/controllers/category_controller.rb:
```diff
@@ -36,8 +39,22 @@ def create
+    @category.descendants.each {|d| d.update_attribute(:gis_enabled, false)}
+    
+    flash[:notice] = "GIS Disabled for #{@category.title}"
+    ajax_redirect_to "#{request.env['HTTP_REFERER']}#category-#{@category.id}"
```
See above comment on path helpers.

benwbrum commented on app/models/article.rb:
```diff
@@ -14,6 +14,10 @@ class Article < ActiveRecord::Base
   validates_presence_of :title
 
+  validates :latitude, allow_blank: true, numericality: { less_than_or_equal_to: 90, greater_than_or_equal_to: -90}
+  validates :longitude, allow_blank: true, numericality: { less_than_or_equal_to: 180, greater_than_or_equal_to: -180}
```
Let's add the following comment to make these fields explicit:
`# latitude and longitude contain modern coordinates for places and are normalized as such.  They do not represent coordinates found within transcribed texts.`

benwbrum replied:
I've just tried this out in the UI, and wonder about the proper behavior when people enter digits longer than the precision allows.

On one hand, we truncate the values silently, which is usually the wrong thing to do.  On the other hand, truncating lat/lon can be absolutely correct for GIS data, in which transformations (division by something close to one, like `3456/3457` may be applied to adjust data from one coordinate system to another) may extend digits of data far beyond what is sane or sensible: We have seen city boundary points expressed with enough precision to identify positions at the sub-atomic level due to these transformations.

I think that if we truncated values to a number of significant digits representing 20m precision or less, and flash that as part of the success message when a user saves, that will be the best compromise.

jackweinbender replied:
1. Is the lat/long comment meant to be for the user (and thus, in the view), or just noted in the model?

2. I completely agree re: truncation. Should we change the precision of the DB table, then, to align with this ~20m precision? I found a handy SO answer which indicates that would mean truncating to 4 decimal places (that's 11m precision), which is about as good as standard consumer GPS does. Or maybe we should make it 5 decimal places just to be a bit conservative? It's currently at 8, which is obviously too precise; I just picked it because Google seemed to use that for Maps (though, I'm counting 7, now).

jackweinbender replied:
I went with 5 digits of precision for the DB table, and added a `GIS_DECIMAL PRECISION` constant to the article controller which governs how it determines if the DB truncated anything as well as how many zeros to pad with. I thought the padding was a nice way to indicate to users what the precision should be (?).

benwbrum commented on app/controllers/category_controller.rb:
```diff
@@ -36,8 +39,22 @@ def create
+    @category.update_attribute(:gis_enabled, true)
+    @category.descendants.each {|d| d.update_attribute(:gis_enabled, true)}
+    
+    flash[:notice] = "GIS Enabled for #{@category.title}"
```
Can we have the success notice mention the children, to make this explicit?  Something like 
`GIS Enabled for #{@category.title} and #{@category.descendants.count} child categories` or `GIS Enabled for #{@category.title} and all children` would make the cascading behavior clearer to users.

benwbrum commented on app/models/article.rb:
```diff
@@ -65,6 +69,10 @@ def self.delete_orphan_articles
   end
+  
+  def gis_enabled?
+    self.categories.any? { |c| c.gis_enabled }
```
Can we rewrite this to push the logic into the database query, rather than needing to loop through results?  `self.categories.where(:gis_enabled => true).present?` might work.

benwbrum commented:
We also need a set of spec tests before we can merge this.  You may be able to find an existing test suite that modifies subjects/articles and categories and add tests for the permutations of gis attribute cascading and lat/lon field display there.

jackweinbender commented:
- [x] Use path helpers
- [ ] Update Lat/Long Fields
- [ ] Add Lat/Long truncation flash message
- [x] Added affected children count to En/Disable GIS action
- [x] Moved `.any?` logic into Db query
- [x] Added tests for GIS enable/disable and conditional dispaly of Lat/Lon fields

saracarl commented:
Is this ready?

jackweinbender commented:
Yes, this is ready. 
Sorry, I wasn't sure how to mark the code review part as "complete."
